include/nbl/video/IGPUCommandPool.h

```diff
@@ -31,13 +66,219 @@ class IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
+    static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+    uint32_t m_size;
+protected:
+    ICommand(uint32_t size) : m_size(size) {}
+};
+
+class alignas(COMMAND_SEGMENT_ALIGNMENT) IGPUCommandPool::CommandSegment
+{
+public:
+    struct Iterator;
+
+    struct params_t
+    {
+        core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+        CommandSegment* m_next;
+    } params;
+
+    static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - sizeof(params_t);
+    alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+    CommandSegment()
+    {
+        params.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, COMMAND_SEGMENT_SIZE);
+        params.m_next = nullptr;
+
+        wipeNextCommandSize();
+    }
+
+    template <typename Cmd, typename... Args>
+    // Cmd* allocate(const Args&... args)
+    Cmd* allocate(Args&&... args)
+    {
+        const uint32_t cmdSize = Cmd::calc_size(args...);
+        const auto address = params.m_commandAllocator.alloc_addr(cmdSize, alignof(Cmd));
+        if (address == decltype(params.m_commandAllocator)::invalid_address)
+            return nullptr;
+
+        wipeNextCommandSize();
+
+        void* cmdMem = m_data + address;
+        return new (cmdMem) Cmd(args...);
+    }
+
+private:
+    void wipeNextCommandSize()
+    {
+        const auto cursor = params.m_commandAllocator.get_allocated_size();
+        const uint32_t wipeSize = offsetof(IGPUCommandPool::ICommand, m_size) + sizeof(IGPUCommandPool::ICommand::m_size);
+        if (cursor + wipeSize < params.m_commandAllocator.get_total_size())
+            memset(m_data + cursor, 0, wipeSize);
+    }
+};
+
+struct IGPUCommandPool::CommandSegment::Iterator
+{
+    CommandSegment* m_segment = nullptr;
+    ICommand* m_cmd = nullptr;
+};
+
+class IGPUCommandPool::CBindIndexBufferCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CBindIndexBufferCmd(const core::smart_refctd_ptr<const video::IGPUBuffer>& indexBuffer) : ICommand(calc_size(indexBuffer)), m_indexBuffer(indexBuffer) {}
+
+    static uint32_t calc_size(const core::smart_refctd_ptr<const video::IGPUBuffer>& indexBuffer)
+    {
+        return core::alignUp(sizeof(void*), alignof(CBindIndexBufferCmd));
+    }
+
+protected:
+    core::smart_refctd_ptr<const video::IGPUBuffer> m_indexBuffer;
+};
+
+class IGPUCommandPool::CDrawCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CDrawCmd() : ICommand(calc_size()) {}
+
+    static uint32_t calc_size()
+    {
+        return core::alignUp(1u, alignof(CDrawCmd));
+    }
+};
+
+class IGPUCommandPool::CDrawIndexedCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CDrawIndexedCmd() : ICommand(calc_size()) {}
+
+    static uint32_t calc_size()
+    {
+        return core::alignUp(1u, alignof(CDrawIndexedCmd));
+    }
+};
```

Comment:
if you accept my Quality of Life improvement, then there's no need for these commands as they don't do anything.

Reply:
Resolved.